modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj

```diff
@@ -801,6 +819,65 @@
   (compile-cond (negate subclause)))
 
 
+;;; ----------------------------------------------------- joins ------------------------------------------------------
+
+(defn- find-source-table-id [join-or-query]
+  (or (some-> join-or-query :source-table)
+      (some-> join-or-query :source-query recur)))
+
+(defn- find-source-collection [join]
+  (or (-> join :source-query :collection)
+      (-> join find-source-table-id qp.store/table :name)))
+
+(defn- localize-join-alias [expr alias]
+  (mbql.u/replace expr
+    [:field id-or-name (m :guard (every-pred map? #(= (:join-alias %) alias)))]
+    [:field id-or-name (set/rename-keys m {:join-alias ::join-local})]))
+
+(declare ^:private mbql->native-rec)
+
+(defn- handle-join [pipeline-ctx
+                    {:keys [alias condition fields source-query strategy] :as join}]
+  (let [result-fields (mapcat #(% source-query) [:fields :breakout :aggregation])
+        {:keys [projections] pipeline :query
+         :or {projections [] pipeline []}}
+        (when source-query
+          (if-let [native (:native source-query)]
+            {:projections (:projections source-query)
+             :query (:query native)}
+            (binding [*field-mappings* (apply dissoc *field-mappings* result-fields)
+                      *query* (assoc (select-keys *query* [:database :type])
+                                     :query source-query)]
+              (mbql->native-rec source-query))))
+        field-mappings (when source-query
+                         (zipmap result-fields
+                                 projections))
+        own-fields (->> (mbql.u/match condition
+                          [:field _ (_ :guard #(not= (:join-alias %) alias))])
+                        (remove *field-mappings*))
+        mapping (map (fn [f] (let [n (str/replace (->lvalue f) "~" "")]
+                               {:field f, :rvalue (->rvalue f), :alias (-> (format "let_%s_" n) gensym name)}))
+                     own-fields)
+        let-map (into {} (map (juxt :alias :rvalue)) mapping)
+        context-map (into {} (map (juxt :field #(str \$ (:alias %)))) mapping)
+        lookup-as (get-join-alias alias)]
+    (binding [*field-mappings* (merge *field-mappings* field-mappings context-map)]
+      (let [pipeline (cond-> pipeline
+                       condition (conj {$match (compile-filter (localize-join-alias condition alias))}))
+            stages (cond-> [{$lookup {:from (find-source-collection join)
+                                      :let let-map
+                                      :pipeline pipeline
+                                      :as lookup-as}}
+                            {$unwind {:path (str \$ lookup-as)
+                                      :preserveNullAndEmptyArrays (= strategy :left-join)}}]
+                     fields (conj))]
+        (-> pipeline-ctx
+            (update :projections into projections)
+            (update :query into stages))))))
```

Comment:
It could be nice to document this a bit so you can get an idea of what's going on without having to start at 40 lines of code for 10 minutes

Not sure if it's possible but breaking it out into a couple of smaller functions could make it more readable too